rust_src/src/lisp.rs

```diff
@@ -576,78 +589,30 @@ impl LispObject {
 /// of arguments.
 pub const MANY: i16 = -2;
 
-/// Internal function to get a displayable string out of a Lisp string.
-fn display_string(obj: LispObject) -> String {
-    let s = obj.as_string().unwrap();
-    let slice = unsafe { slice::from_raw_parts(s.const_data_ptr(), s.len_bytes() as usize) };
-    String::from_utf8_lossy(slice).into_owned()
-}
-
 impl Debug for LispObject {
     fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
         let ty = self.get_type();
-        let self_ptr = &self as *const _ as usize;
+        let self_ptr = &self as *const _;
         if ty as u8 >= 8 {
-            write!(
+            return write!(
                 f,
-                "#<INVALID-OBJECT @ {:#X}: VAL({:#X})>",
+                "#<INVALID-OBJECT @ {:p}: VAL({:#X})>",
                 self_ptr,
                 self.to_C()
-            )?;
-            return Ok(());
+            );
         }
         if self.is_nil() {
             return write!(f, "nil");
         }
         match ty {
-            Lisp_Type::Lisp_Symbol => {
-                let name = self.as_symbol_or_error().symbol_name();
-                write!(f, "'{}", display_string(name))?;
-            }
-            Lisp_Type::Lisp_Cons => {
-                let mut cdr = *self;
-                write!(f, "'(")?;
-                while let Some(cons) = cdr.as_cons() {
-                    write!(f, "{:?} ", cons.car())?;
-                    cdr = cons.cdr();
-                }
-                if cdr.is_nil() {
-                    write!(f, ")")?;
-                } else {
-                    write!(f, ". {:?}", cdr)?;
-                }
-            }
-            Lisp_Type::Lisp_Float => {
-                write!(f, "{}", self.as_float().unwrap())?;
-            }
-            Lisp_Type::Lisp_Vectorlike => {
-                let vl = self.as_vectorlike().unwrap();
-                if vl.is_vector() {
-                    write!(f, "[")?;
-                    for el in vl.as_vector().unwrap().as_slice() {
-                        write!(f, "{:?} ", el)?;
-                    }
-                    write!(f, "]")?;
-                } else {
-                    write!(
-                        f,
-                        "#<VECTOR-LIKE @ {:#X}: VAL({:#X})>",
-                        self_ptr,
-                        self.to_C()
-                    )?;
-                }
-            }
-            Lisp_Type::Lisp_Int0 | Lisp_Type::Lisp_Int1 => {
-                write!(f, "{}", self.as_fixnum().unwrap())?;
-            }
-            Lisp_Type::Lisp_Misc => {
-                write!(f, "#<MISC @ {:#X}: VAL({:#X})>", self_ptr, self.to_C())?;
-            }
-            Lisp_Type::Lisp_String => {
-                write!(f, "{:?}", display_string(*self))?;
-            }
+            Lisp_Type::Lisp_Symbol => write!(f, "{:?}", self.force_symbol()),
```

Comment:
Why are there all these separate `force_whatever` functions? It seems like it would be easier to name them all `force` or `force_type` or something, and some of these branches could be consolidated.

Reply:
They are all members of LispObject so they need different names.

The intent here is to provide the equivalent of XINT or XCONS. These functions do no checks and act like cast operators.

Reply:
They are also meant to be fast / least effort. Having a if/else or match logic would defeat that purpose. These would replace some of the `as_foo_or_error()` calls when we know for certain the type of object we have. Like, we know we just made a CONS cell no need to wrap it an Option and unwrap it using `as_cons_or_error()`.

Reply:
I could have done this as two PRs. But I thought the existing type `match` was the perfect reason to add the `force_` functions. The alternative is like I did in the `internal_equal` PR:

```
if let (Some(a), Some(b)) = (thing1.as_foo(), thing2.as_foo()) {
    ...
} else if let (Some(a), Some(b)) = (thing1.as_bar(), thing2.as_bar()) {
    ...
}
```

But the direct type and force is way more efficient.

Reply:
I'd be fine with `force_foo` but we could also go with `as_foo_unchecked` which is more discoverable since we already have several `as_foo` and `as_foo_or_error`. Given that the `XFOO` macros are used a lot, same might hold in Rust code (?) which would make a short name more appealing.

Reply:
I chose `force_foo` because it was shorter than `as_foo_unchecked`. I expect these to be used frequently. Another option would be macros like `CONS!()` or `FIXNUM!()` which evaluate to something like `obj.as_foo().unwrap()` or `Option<LispFoo>::from(obj).unwrap()` internally.